WrapperDownloader should retry on Layer3/Layer4 network errors

### Description

WrapperDownloader became more stable in CI after #11766 added Layer7 retries for HTTP 5xx.

But I think we should also retry on lower-level network failures (e.g. DNS failure, connect timeout, read timeout, etc). I already encountered this from github actions:
```
Run ./gradlew localSettings --max-workers 2
  ./gradlew localSettings --max-workers 2
  shell: /usr/bin/bash -e {0}
  env:
    JAVA_HOME: /opt/hostedtoolcache/Java_Temurin-Hotspot_jdk/17.0.4-1/x64
Downloading gradle-wrapper.jar from https://raw.githubusercontent.com/gradle/gradle/v7.3.3/gradle/wrapper/gradle-wrapper.jar
ERROR: Could not download gradle-wrapper.jar (ConnectException: Connection timed out).
Error: Process completed with exit code 3.
```

--- a/buildSrc/src/main/java/org/apache/lucene/gradle/WrapperDownloader.java
+++ b/buildSrc/src/main/java/org/apache/lucene/gradle/WrapperDownloader.java
@@ -103,7 +103,16 @@ public class WrapperDownloader {
       HttpURLConnection connection;
       while (true) {
         connection = (HttpURLConnection) url.openConnection();
-        connection.connect();
+        try {
+          connection.connect();
+        } catch (IOException e) {
+          if (retries-- > 0) {
+            // Retry after a short delay
+            System.err.println("Error connecting to server: " + e + ", will retry in " + retryDelay + " seconds.");
+            Thread.sleep(TimeUnit.SECONDS.toMillis(retryDelay));
+            continue;
+          }
+        }
 
         switch (connection.getResponseCode()) {
           case HttpURLConnection.HTTP_INTERNAL_ERROR:

